File: web/src/mocks/slos.ts
// MSW handlers for the SLO endpoints (internal/api/slos.go, docs/contracts/slo.md): two objectives of the
// mock shop — checkout availability (28 days, budget nearly gone, fast window burning) and catalog latency
// (7 days, healthy) — with in-memory CRUD.
import { http, HttpResponse, type HttpResponseResolver } from "msw";
import type { Slo, SloBudget, SloInput, SloPoint } from "@/api/slos";
import { authenticate } from "./account";
import { formatTs } from "./fixtures";

const API = "*/api/v1";

type Code = "invalid_argument" | "not_found" | "permission_denied" | "failed_precondition";
const STATUS: Record<Code, number> = { invalid_argument: 400, not_found: 404, permission_denied: 403, failed_precondition: 409 };
const fail = (code: Code, message: string) => HttpResponse.json({ error: { code, message } }, { status: STATUS[code] });

function authed(resolver: HttpResponseResolver): HttpResponseResolver {
  return (info) => {
    const ctx = authenticate(info.request);
    if (ctx instanceof Response) return ctx;
    return resolver(info);
  };
}

/** Writes need a signed-in member or higher (slo.md §1). */
function writer(request: Request): Response | null {
  const ctx = authenticate(request);
  if (ctx instanceof Response) return ctx;
  if (ctx.kind !== "session") return fail("permission_denied", "this operation requires a signed-in user; API keys are read-only");
  if (ctx.role === "viewer") return fail("permission_denied", `your role (${ctx.role}) does not allow this operation`);
  return null;
}

const SLO_IDS = {
  availability: "51000000-0000-4000-8000-000000000001",
  latency: "51000000-0000-4000-8000-000000000002",
};

/** Test helper: the seeded objectives (availability burning, latency healthy). */
export const MOCK_SLO_IDS = SLO_IDS;

/** Bad request share per SLO, so the mock budgets are deterministic. */
const BAD_RATIO: Record<string, number> = { [SLO_IDS.availability]: 0.0012, [SLO_IDS.latency]: 0.004 };

function seed(): Slo[] {
  const now = Date.now();
  const base = { created_by_email: "admin@openlog.local", updated_by_email: "admin@openlog.local", created_at: formatTs(now - 30 * 86_400_000), updated_at: formatTs(now - 86_400_000) };
  return [
    {
      id: SLO_IDS.availability, name: "Checkout availability", description: "Successful checkout requests", service_name: "frontend",
      service_namespace: "shop", environment: "prod", sli_type: "availability", latency_threshold_ms: null, objective: 99.9,
      window_days: 28, ...base,
    },
    {
      id: SLO_IDS.latency, name: "Catalog latency", description: "", service_name: "catalog", service_namespace: null,
      environment: "prod", sli_type: "latency", latency_threshold_ms: 300, objective: 99, window_days: 7, ...base,
    },
  ];
}

let slos: Slo[] = seed();
let nextId = 3;

/** Test helper: forget created SLOs. */
export function resetMockSlos(): void {
  slos = seed();
  nextId = 3;
}

function budget(requests: number, badRatio: number, objective: number): SloBudget {
  const bad = Math.round(requests * badRatio);
  const good = requests - bad;
  const allowed = 1 - objective / 100;
  const budgetRequests = allowed * requests;
  return {
    requests, good, bad, sli: requests ? good / requests : null,
    budget_requests: budgetRequests, budget_consumed: bad, budget_remaining: budgetRequests - bad,
    remaining_ratio: requests ? 1 - bad / budgetRequests : null,
    burn_rate: requests ? bad / requests / allowed : null,
    met: requests === 0 || good / requests >= objective / 100,
  };
}

function status(slo: Slo) {
  const now = Date.now();
  const minutes = slo.window_days * 24 * 60;
  const requests = minutes * 200;
  return {
    from: formatTs(now - slo.window_days * 86_400_000),
    to: formatTs(now),
    window_days: slo.window_days,
    budget: budget(requests, BAD_RATIO[slo.id] ?? 0.0005, slo.objective),
  };
}

function series(slo: Slo, points = 60): SloPoint[] {
  const now = Date.now();
  const stepMs = Math.round((slo.window_days * 86_400_000) / points);
  const total = status(slo).budget;
  const out: SloPoint[] = [];
  let consumed = 0;
  for (let i = 0; i < points; i++) {
    const requests = total.requests / points;
    // The budget burns faster in the last fifth of the window (a recent incident).
    const share = i > points * 0.8 ? 3 : 0.5;
    const bad = Math.min(total.bad - consumed, (total.bad / points) * share);
    consumed += Math.max(0, bad);
    out.push({
      t: now - (points - i) * stepMs,
      requests,
      good: requests - Math.max(0, bad),
      bad: Math.max(0, bad),
      sli: requests ? (requests - Math.max(0, bad)) / requests : null,
      burn_rate: requests ? Math.max(0, bad) / requests / (1 - slo.objective / 100) : null,
      remaining_ratio: total.budget_requests ? 1 - consumed / total.budget_requests : null,
    });
  }
  return out;
}

function burn(slo: Slo) {
  const breaching = slo.id === SLO_IDS.availability;
  const window = (name: string, factor: number, long: number, short: number, rate: number | null) => {
    const requests = (long / 60) * 200;
    const badRatio = rate === null ? 0 : rate * (1 - slo.objective / 100);
    return {
      name, factor, long_seconds: long, short_seconds: short,
      long: budget(requests, badRatio, slo.objective),
      short: budget((short / 60) * 200, badRatio, slo.objective),
      rate, ratio: rate === null ? null : rate / factor, breaching: rate !== null && rate >= factor,
    };
  };
  return [
    window("fast", 14.4, 3600, 300, breaching ? 18.2 : 0.4),
    window("slow", 6, 21600, 1800, breaching ? 4.1 : 0.3),
  ];
}

function validate(body: Partial<SloInput>): string | null {
  if (!body.name || typeof body.name !== "string" || body.name.trim() === "") return "name: must be 1-200 characters";
  if (!body.service_name || typeof body.service_name !== "string") return "service_name: required, at most 512 bytes";
  if (body.sli_type !== "availability" && body.sli_type !== "latency") return "sli_type: must be availability or latency";
  if (body.sli_type === "latency" && !(Number.isInteger(body.latency_threshold_ms) && (body.latency_threshold_ms as number) >= 1)) {
    return "latency_threshold_ms: must be a whole number of milliseconds between 1 and 600000";
  }
  if (typeof body.objective !== "number" || body.objective < 50 || body.objective >= 100) {
    return "objective: must be at least 50 and below 100 (percent)";
  }
  if (![7, 28, 30].includes(Number(body.window_days))) return "window_days: must be 7, 28 or 30";
  return null;
}

function stored(body: SloInput, existing?: Slo): Slo {
  const now = Date.now();
  return {
    id: existing?.id ?? `51000000-0000-4000-8000-${String(nextId++).padStart(12, "0")}`,
    name: body.name.trim(),
    description: body.description ?? "",
    service_name: body.service_name.trim(),
    service_namespace: body.service_namespace ?? null,
    environment: body.environment ?? null,
    sli_type: body.sli_type,
    latency_threshold_ms: body.sli_type === "latency" ? (body.latency_threshold_ms ?? null) : null,
    objective: body.objective,
    window_days: body.window_days,
    created_by_email: existing?.created_by_email ?? "admin@openlog.local",
    updated_by_email: "admin@openlog.local",
    created_at: existing?.created_at ?? formatTs(now),
    updated_at: formatTs(now),
  };
}

export const sloHandlers = [
  http.get(`${API}/slos`, authed(({ request }) => {
    const withStatus = new URL(request.url).searchParams.get("status") !== "false";
    return HttpResponse.json({
      slos: slos.map((s) => ({ ...s, status: withStatus ? status(s) : null })),
      status_truncated: false,
    });
  })),

  http.post(`${API}/slos`, async ({ request }) => {
    const denied = writer(request);
    if (denied) return denied;
    const body = (await request.json().catch(() => ({}))) as Partial<SloInput>;
    const err = validate(body);
    if (err) return fail("invalid_argument", err);
    const s = stored(body as SloInput);
    slos = [...slos, s];
    return HttpResponse.json(s, { status: 201 });
  }),

  http.get(`${API}/slos/:id`, authed(({ params }) => {
    const s = slos.find((x) => x.id === params.id);
    return s ? HttpResponse.json(s) : fail("not_found", "slo not found");
  })),

  http.put(`${API}/slos/:id`, async ({ request, params }) => {
    const denied = writer(request);
    if (denied) return denied;
    const existing = slos.find((x) => x.id === params.id);
    if (!existing) return fail("not_found", "slo not found");
    const body = (await request.json().catch(() => ({}))) as Partial<SloInput>;
    const err = validate(body);
    if (err) return fail("invalid_argument", err);
    const s = stored(body as SloInput, existing);
    slos = slos.map((x) => (x.id === s.id ? s : x));
    return HttpResponse.json(s);
  }),

  http.delete(`${API}/slos/:id`, ({ request, params }) => {
    const denied = writer(request);
    if (denied) return denied;
    if (!slos.some((x) => x.id === params.id)) return fail("not_found", "slo not found");
    slos = slos.filter((x) => x.id !== params.id);
    return new HttpResponse(null, { status: 204 });
  }),

  http.get(`${API}/slos/:id/results`, authed(({ params }) => {
    const s = slos.find((x) => x.id === params.id);
    if (!s) return fail("not_found", "slo not found");
    return HttpResponse.json({ slo: s, status: status(s), step: "600s", burn: burn(s), series: series(s) });
  })),
];

File: internal/api/slos.go
package api

import (
	"errors"
	"fmt"
	"net/http"
	"strconv"
	"time"

	"github.com/onuragtas/openlog/internal/api/query"
	"github.com/onuragtas/openlog/internal/auth"
	"github.com/onuragtas/openlog/internal/slo"
)

// Service level objectives (docs/contracts/slo.md, api.md "Service level objectives"). Definitions live in
// PostgreSQL; the SLI, the error budget and the burn rates are computed at query time from
// apm_transactions_1m through the tenant-scoped query layer (nothing is precomputed).
//
// Reads are telemetry reads (any role, API keys too); writes need a signed-in member or higher and are
// audited by the store (slo.create/update/delete).

// sloStatusLimit bounds how many SLOs of a list response get their current budget computed (one small
// aggregate query each); the rest are returned without a status.
const sloStatusLimit = 50

// SetSLOs enables the SLO endpoints (postgres auth mode). Must be called before Run.
func (s *Server) SetSLOs(store slo.Store) {
	s.slos = store
	s.srv.Handler = s.Handler()
}

func (s *Server) sloRoutes(mux *http.ServeMux) {
	if s.slos == nil {
		return
	}
	route := func(pattern string, h handlerFunc) { mux.Handle(pattern, s.wrap(pattern, h)) }
	route("GET /api/v1/slos", s.listSLOs)
	route("GET /api/v1/slos/{id}", s.getSLO)
	route("GET /api/v1/slos/{id}/results", s.sloResults)
	if s.accounts != nil {
		s.sloWriteRoute(mux, "POST /api/v1/slos", s.createSLO)
		s.sloWriteRoute(mux, "PUT /api/v1/slos/{id}", s.updateSLO)
		s.sloWriteRoute(mux, "DELETE /api/v1/slos/{id}", s.deleteSLO)
	}
}

type sloWriteHandler func(w http.ResponseWriter, r *http.Request, p *auth.Principal) error

// sloWriteRoute registers a mutating endpoint: wrap authenticates (CSRF included), the handler additionally
// needs a principal whose role may write SLOs (a member, or an API key with that role).
func (s *Server) sloWriteRoute(mux *http.ServeMux, pattern string, h sloWriteHandler) {
	mux.Handle(pattern, s.wrap(pattern, func(w http.ResponseWriter, r *http.Request, _ *query.Scope) error {
		noStore(w)
		p, _ := auth.PrincipalFrom(r.Context())
		if ae := authorize(p, auth.ActWriteSLOs); ae != nil {
			return ae
		}
		return sloError(h(w, r, p))
	}))
}

// sloError maps store and validation errors to the API error shape.
func sloError(err error) error {
	var ve *slo.ValidationError
	switch {
	case errors.As(err, &ve):
		return &apiError{http.StatusBadRequest, "invalid_argument", ve.Error()}
	case errors.Is(err, slo.ErrNotFound):
		return notFound("slo not found")
	case errors.Is(err, slo.ErrLimit):
		return &apiError{http.StatusConflict, "failed_precondition",
			fmt.Sprintf("the organization already has the maximum number of SLOs (%d); delete some first", slo.MaxPerOrg)}
	}
	return err
}

// ---- response shapes ----

type sloJSON struct {
	ID                 string   `json:"id"`
	Name               string   `json:"name"`
	Description        string   `json:"description"`
	ServiceName        string   `json:"service_name"`
	ServiceNamespace   *string  `json:"service_namespace"`
	Environment        *string  `json:"environment"`
	SLIType            string   `json:"sli_type"`
	LatencyThresholdMs *float64 `json:"latency_threshold_ms"`
	Objective          float64  `json:"objective"`
	WindowDays         int      `json:"window_days"`
	CreatedByEmail     string   `json:"created_by_email"`
	UpdatedByEmail     string   `json:"updated_by_email"`
	CreatedAt          string   `json:"created_at"`
	UpdatedAt          string   `json:"updated_at"`
}

func sloResponse(x *slo.SLO) sloJSON {
	out := sloJSON{ID: x.ID, Name: x.Name, Description: x.Description, ServiceName: x.ServiceName,
		ServiceNamespace: x.ServiceNamespace, Environment: x.Environment, SLIType: x.SLIType,
		Objective: x.Objective, WindowDays: x.WindowDays, CreatedByEmail: x.CreatedByEmail,
		UpdatedByEmail: x.UpdatedByEmail, CreatedAt: formatTime(x.CreatedAt), UpdatedAt: formatTime(x.UpdatedAt)}
	if x.SLIType == slo.SLILatency {
		out.LatencyThresholdMs = optFloat(x.LatencyThresholdMs)
	}
	return out
}

// sloBudgetJSON is the error budget over one range (slo.md §2). Ratios are null without requests.
type sloBudgetJSON struct {
	Requests        float64  `json:"requests"`
	Good            float64  `json:"good"`
	Bad             float64  `json:"bad"`
	SLI             *float64 `json:"sli"`
	BudgetRequests  float64  `json:"budget_requests"`
	BudgetConsumed  float64  `json:"budget_consumed"`
	BudgetRemaining float64  `json:"budget_remaining"`
	RemainingRatio  *float64 `json:"remaining_ratio"`
	BurnRate        *float64 `json:"burn_rate"`
	Met             bool     `json:"met"`
}

func budgetResponse(b slo.Budget) sloBudgetJSON {
	return sloBudgetJSON{Requests: b.Requests, Good: b.Good, Bad: b.Bad, SLI: optFloat(b.SLI),
		BudgetRequests: b.BudgetRequests, BudgetConsumed: b.BudgetConsumed, BudgetRemaining: b.BudgetRemaining,
		RemainingRatio: optFloat(b.RemainingRatio), BurnRate: optFloat(b.BurnRate), Met: b.Met()}
}

type sloStatusJSON struct {
	From       string        `json:"from"`
	To         string        `json:"to"`
	WindowDays int           `json:"window_days"`
	Budget     sloBudgetJSON `json:"budget"`
}

type sloListItemJSON struct {
	sloJSON
	Status *sloStatusJSON `json:"status"`
}

// ---- CRUD ----

func (s *Server) listSLOs(w http.ResponseWriter, r *http.Request, sc *query.Scope) error {
	noStore(w)
	p, _ := auth.PrincipalFrom(r.Context())
	list, err := s.slos.List(r.Context(), p.OrgID)
	if err != nil {
		return sloError(err)
	}
	status := true
	if v := r.URL.Query().Get("status"); v != "" {
		if status, err = strconv.ParseBool(v); err != nil {
			return badRequest("status must be true or false")
		}
	}
	out := make([]sloListItemJSON, 0, len(list))
	truncated := false
	for i := range list {
		item := sloListItemJSON{sloJSON: sloResponse(&list[i])}
		switch {
		case !status:
		case i < sloStatusLimit:
			st, err := s.sloStatus(r, sc, &list[i], list[i].Window())
			if err != nil {
				return err
			}
			item.Status = st
		default:
			truncated = true
		}
		out = append(out, item)
	}
	writeJSON(w, http.StatusOK, map[string]any{"slos": out, "status_truncated": truncated})
	return nil
}

// sloStatus computes the budget over the rolling window ending at the last complete minute with a single
// bucket (sums and merged histograms are exact whatever the step is, slo.md §2).
func (s *Server) sloStatus(r *http.Request, sc *query.Scope, x *slo.SLO, window time.Duration) (*sloStatusJSON, error) {
	to := s.now().UTC().Truncate(time.Minute)
	from := to.Add(-window)
	buckets, err := slo.Load(r.Context(), sc, x, from, to, window)
	if err != nil {
		return nil, err
	}
	b := slo.Compute(slo.Sum(buckets), x.ObjectiveFraction())
	return &sloStatusJSON{From: formatTime(from), To: formatTime(to), WindowDays: x.WindowDays, Budget: budgetResponse(b)}, nil
}

func (s *Server) sloByID(r *http.Request) (*slo.SLO, error) {
	p, _ := auth.PrincipalFrom(r.Context())
	x, err := s.slos.Get(r.Context(), p.OrgID, r.PathValue("id"))
	if err != nil {
		return nil, sloError(err)
	}
	return x, nil
}

func (s *Server) getSLO(w http.ResponseWriter, r *http.Request, _ *query.Scope) error {
	noStore(w)
	x, err := s.sloByID(r)
	if err != nil {
		return err
	}
	writeJSON(w, http.StatusOK, sloResponse(x))
	return nil
}

func decodeSLOInput(r *http.Request) (slo.Input, error) {
	var in slo.Input
	if err := decodeJSON(r, &in); err != nil {
		return in, err
	}
	return in, in.Validate()
}

func (s *Server) sloActor(r *http.Request, p *auth.Principal) slo.Actor {
	return slo.Actor{UserID: p.UserID, Email: p.Email, IP: s.accounts.Meta(r).IP}
}

func (s *Server) createSLO(w http.ResponseWriter, r *http.Request, p *auth.Principal) error {
	in, err := decodeSLOInput(r)
	if err != nil {
		return err
	}
	x, err := s.slos.Create(r.Context(), p.OrgID, in, s.sloActor(r, p))
	if err != nil {
		return err
	}
	writeJSON(w, http.StatusCreated, sloResponse(x))
	return nil
}

func (s *Server) updateSLO(w http.ResponseWriter, r *http.Request, p *auth.Principal) error {
	in, err := decodeSLOInput(r)
	if err != nil {
		return err
	}
	x, err := s.slos.Update(r.Context(), p.OrgID, r.PathValue("id"), in, s.sloActor(r, p))
	if err != nil {
		return err
	}
	writeJSON(w, http.StatusOK, sloResponse(x))
	return nil
}

func (s *Server) deleteSLO(w http.ResponseWriter, r *http.Request, p *auth.Principal) error {
	if err := s.slos.Delete(r.Context(), p.OrgID, r.PathValue("id"), s.sloActor(r, p)); err != nil {
		return err
	}
	w.WriteHeader(http.StatusNoContent)
	return nil
}

// ---- results ----

type sloBurnJSON struct {
	Name         string        `json:"name"`
	Factor       float64       `json:"factor"`
	LongSeconds  int           `json:"long_seconds"`
	ShortSeconds int           `json:"short_seconds"`
	Long         sloBudgetJSON `json:"long"`
	Short        sloBudgetJSON `json:"short"`
	Rate         *float64      `json:"rate"`
	Ratio        *float64      `json:"ratio"`
	Breaching    bool          `json:"breaching"`
}

type sloPointJSON struct {
	T              int64    `json:"t"`
	Requests       float64  `json:"requests"`
	Good           float64  `json:"good"`
	Bad            float64  `json:"bad"`
	SLI            *float64 `json:"sli"`
	BurnRate       *float64 `json:"burn_rate"`
	RemainingRatio *float64 `json:"remaining_ratio"`
}

// sloResults serves GET /slos/{id}/results: the budget over the SLO's rolling window, the burn-rate windows
// of the slo_burn defaults and the burndown series (one point per step).
func (s *Server) sloResults(w http.ResponseWriter, r *http.Request, sc *query.Scope) error {
	noStore(w)
	x, err := s.sloByID(r)
	if err != nil {
		return err
	}
	to := s.now().UTC().Truncate(time.Minute)
	from := to.Add(-x.Window())
	step, err := apmStep(r, from, to)
	if err != nil {
		return err
	}
	buckets, err := slo.Load(r.Context(), sc, x, from, to, step)
	if err != nil {
		return err
	}
	objective := x.ObjectiveFraction()
	status := sloStatusJSON{From: formatTime(from), To: formatTime(to), WindowDays: x.WindowDays,
		Budget: budgetResponse(slo.Compute(slo.Sum(buckets), objective))}

	burnBuckets, err := slo.LoadBurn(r.Context(), sc, x, to, slo.DefaultBurnWindows)
	if err != nil {
		return err
	}
	burn := make([]sloBurnJSON, 0, len(slo.DefaultBurnWindows))
	for _, bw := range slo.DefaultBurnWindows {
		b := slo.BurnAt(burnBuckets, slo.BurnStep, to, objective, bw)
		burn = append(burn, sloBurnJSON{Name: bw.Name, Factor: bw.Factor, LongSeconds: int(bw.Long / time.Second),
			ShortSeconds: int(bw.Short / time.Second), Long: budgetResponse(b.Long), Short: budgetResponse(b.Short),
			Rate: optFloat(b.Rate), Ratio: optFloat(b.Ratio), Breaching: b.Breaching()})
	}

	points := slo.Series(buckets, objective)
	series := make([]sloPointJSON, 0, len(points))
	for _, p := range points {
		series = append(series, sloPointJSON{T: p.T.UnixMilli(), Requests: p.Requests, Good: p.Good, Bad: p.Bad,
			SLI: optFloat(p.SLI), BurnRate: optFloat(p.BurnRate), RemainingRatio: optFloat(p.RemainingRatio)})
	}
	writeJSON(w, http.StatusOK, map[string]any{"slo": sloResponse(x), "status": status, "step": formatStep(step),
		"burn": burn, "series": series})
	return nil
}
